node/pkg/watchers/near/finalizer.go

```diff
@@ -0,0 +1,128 @@
+	if err := nearapi.IsWellFormedHash(blockHash); err != nil {
+		// SECURITY defense-in-depth: check if block hash is well-formed
+		logger.Error("blockHash invalid", zap.String("error_type", "invalid_hash"), zap.String("blockHash", blockHash), zap.Error(err))
+		return false, nearapi.BlockHeader{}
+	}
+
+	if b, ok := f.finalizedBlocksCache.Get(blockHash); ok {
+		blockHeader := b.(nearapi.BlockHeader)
+		// SECURITY In blocks < 74473147 message timestamps were computed differently and we don't want to re-observe these messages
+		if !f.mainnet || blockHeader.Height > 74473147 {
+			return true, blockHeader
+		}
+	}
+
+	return false, nearapi.BlockHeader{}
+}
+
+// isFinalized() checks if a block is finalized by looking at the local cache first. If therer is an error during execution it returns false.
+// If it is not found in the cache, we walk forward up to nearBlockchainMaxGaps blocks by height,
+// starting at the block's height+2 and check if their value of "last_final_block" matches
+// the block in question.
+// we start at height+2 because NEAR consensus takes at least two blocks to reach finality.
+func (f Finalizer) isFinalized(logger *zap.Logger, ctx context.Context, queriedBlockHash string) (bool, nearapi.BlockHeader) {
+
+	logger.Debug("checking block finalization", zap.String("method", "isFinalized"), zap.String("parameters", queriedBlockHash))
+
+	// check cache first
+	if ok, block := f.isFinalizedCached(logger, ctx, queriedBlockHash); ok {
+		return true, block
+	}
+
+	logger.Debug("block finalization cache miss", zap.String("method", "isFinalized"), zap.String("parameters", queriedBlockHash))
+	f.eventChan <- EVENT_FINALIZED_CACHE_MISS
+
+	queriedBlock, err := f.nearAPI.GetBlock(ctx, queriedBlockHash)
+	if err != nil {
+		return false, nearapi.BlockHeader{}
+	}
+	startingBlockHeight := queriedBlock.Header.Height
+
+	for i := 0; i < nearBlockchainMaxGaps; i++ {
+		block, err := f.nearAPI.GetBlockByHeight(ctx, startingBlockHeight+uint64(2+i))
+		if err != nil {
+			break
+		}
+
+		// SECURITY defense-in-depth check
+		if block.Header.Height != startingBlockHeight+uint64(2+i) {
+			// SECURITY violation: Block height is different than what we queried for
+			logger.Panic("NEAR RPC Inconsistent", zap.String("error_type", "nearapi_inconsistent"), zap.String("inconsistency", "block_height_result_different_from_query"))
+		}
+
+		someFinalBlockHash := block.Header.LastFinalBlock
+
+		// SECURITY defense-in-depth check
+		if someFinalBlockHash == "" || block.Header.Height == 0 || block.Header.Timestamp == 0 {
+			break
+		}
+
+		if queriedBlockHash == someFinalBlockHash {
+			f.setFinalized(logger, ctx, queriedBlock.Header)
+			// block was marked as finalized in the cache, so this should succeed now.
+			// We don't return directly because setFinalized() contains some sanity checks.
+			return f.isFinalizedCached(logger, ctx, queriedBlockHash)
```

Comment:
If the block fails the sanity check in isFinalizedCached(), wouldn't we be stuck in a loop here?

Reply:
no, it'd return that error then, which is what I intended. This way we ensure that all the sanity checks from isFinalizedCached() are also done when calling isFinalized().